### PR TITLE
Plugin utils.

### PR DESCRIPTION
**Pull request by o0Ignition0o**
This commit adds plugin utilities that will allow users to craft Request and Responses, and MockServices so they will be able to test their Layers and Plugins in isolation.

### CODE REVIEW

File: apollo-router-core/src/services/router_service.rs [hunk 1]
```diff
@@ -76,6 +76,21 @@ where
...
 
+        let query = &request.http_request.body().query;
+
+        if query.is_empty() {
+            let res = RouterResponseBuilder::new()
+                .push_error(crate::Error {
+                    message: "Must provide query string.".to_string(),
+                    locations: Default::default(),
+                    path: Default::default(),
+                    extensions: Default::default(),
```
Comment (bnjjj) [inline apollo-router-core/src/services/router_service.rs:87]: nit:
```suggestion
                    ..Default::default()
```

File: apollo-router-core/src/plugin_utils/structures/router_response.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                        errors: this.errors,
+                        extensions: this.extensions.unwrap_or_default(),
+                    }
+                    .into(),
+                )
+                .expect("crate::Response implements Serialize; qed")
+                .into(),
+            context: this.context.unwrap_or_else(|| {
+                Context::new().with_request(Arc::new(
+                    Request::new(crate::Request {
```
Comment (bnjjj) [inline apollo-router-core/src/plugin_utils/structures/router_response.rs:48]: Maybe we could derive `Default` on `Request`

File: apollo-router-core/Cargo.toml [hunk 3]
```diff
@@ -23,21 +24,24 @@ http = "0.2.6"
...
-serde_json_bytes = { version = "0.2.0", features = ["preserve_order"]}
+serde_json_bytes = { version = "0.2.0", features = ["preserve_order"] }
 static_assertions = "1.1.0"
 thiserror = "1.0.30"
 tokio = { version = "1.16.1", features = ["rt", "sync"] }
-tower = { version="0.4.11", features = ["util", "buffer"] }
+tower = { version = "0.4.11", features = ["util", "buffer"] }
 tower-service = "0.3.1"
-tracing = "0.1.29"
+tower-test = { version = "0.4.0", optional = true }
```
Comment (BrynCooke) [inline apollo-router-core/Cargo.toml:43]: Does `tower-test` need to be gated behind the service mock feature?

**Review by BrynCooke — APPROVED** [review]

File: apollo-router-core/Cargo.toml (hunk 3 above)
Comment (o0Ignition0o) [reply apollo-router-core/Cargo.toml:43]: I gated the mock service implementation behind service_mock, but i m not quite sure it adds "this much bloat", should I rather remove the gate allotgether?

File: apollo-router-core/src/services/router_service.rs (hunk 1 above)
Comment (o0Ignition0o) [reply apollo-router-core/src/services/router_service.rs:87]: true! nice catch :D

File: apollo-router-core/src/plugin_utils/structures/router_response.rs (hunk 2 above)
Comment (o0Ignition0o) [reply apollo-router-core/src/plugin_utils/structures/router_response.rs:48]: We could! let's open a followup issue so we can all agree on that

File: apollo-router-core/src/services/router_service.rs (hunk 1 above)
Comment (o0Ignition0o) [reply apollo-router-core/src/services/router_service.rs]: fixed in 4b3cfe2

File: apollo-router-core/src/plugin_utils/structures/router_response.rs (hunk 2 above)
Comment (o0Ignition0o) [reply apollo-router-core/src/plugin_utils/structures/router_response.rs:48]: followed up on #437

File: apollo-router-core/Cargo.toml (hunk 3 above)
Comment (BrynCooke) [reply apollo-router-core/Cargo.toml:43]: Probably a good idea, users will want to test the services/layers that they create and gating the test utils will make things more complicated for them.

File: apollo-router-core/src/plugin_utils/structures/router_request.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
+use std::sync::Arc;
+
+use http::Request;
+use serde_json_bytes::{ByteString, Value};
+
+use crate::{Context, Object};
+
+#[derive(Default, Clone)]
+pub struct RouterRequestBuilder {
```
Comment (garypen) [inline apollo-router-core/src/plugin_utils/structures/router_request.rs:9]: I know we use `TypedBuilder` elsewhere. Could we use it here to reduce boilerplate?

File: apollo-router-core/Cargo.toml (hunk 3 above)
Comment (o0Ignition0o) [reply apollo-router-core/Cargo.toml]: fixed in adc457e

**Review by garypen — COMMENTED** [review]
Looks nice. Just one question about perhaps avoiding hand-coding the builders.

File: apollo-router-core/src/plugin_utils/structures/router_request.rs (hunk 4 above)
Comment (o0Ignition0o) [reply apollo-router-core/src/plugin_utils/structures/router_request.rs:9]: oh indeed, sorry I think you mentioned that before and i totally forgot to try to adress it, lemme check!

Comment (o0Ignition0o) [reply apollo-router-core/src/plugin_utils/structures/router_request.rs]: fixed in d9fcb56

File: apollo-router-core/src/plugin_utils/structures/router_response.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+use crate::{Context, Error, Object, Path};
+use http::StatusCode;
+use http::{Request, Response};
+use serde_json_bytes::{ByteString, Value};
+
+type CompatRequest = Arc<crate::http_compat::Request<crate::Request>>;
+
+#[derive(Default, Clone)]
+pub struct RouterResponseBuilder {
+    label: Option<String>,
```
[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugin_utils/structures/router_response.rs]: fixed in d9fcb56

**Pull request merged**